tests/core/contracts/test_contract_estimateGas.py

```diff
@@ -92,3 +109,31 @@ def test_contract_estimateGas_with_arguments(web3, math_contract, estimateGas, t
     gas_used = txn_receipt.get('gasUsed')
 
     assert abs(gas_estimate - gas_used) < 21000
+
+
+def test_estimateGas_not_sending_ether_to_nonpayable_function(
+        web3,
+        payable_tester_contract,
+        estimateGas,
+        transact):
+    gas_estimate = estimateGas(contract=payable_tester_contract,
+                               contract_function='doNoValueCall')
+
+    txn_hash = transact(
+        contract=payable_tester_contract,
+        contract_function='doNoValueCall')
+
+    txn_receipt = web3.eth.waitForTransactionReceipt(txn_hash)
+    gas_used = txn_receipt.get('gasUsed')
+
+    assert abs(gas_estimate - gas_used) < 21000
+
+
+def test_estimateGas_sending_ether_to_nonpayable_function(
+        web3,
+        payable_tester_contract,
+        estimateGas):
+    with pytest.raises(ValidationError):
+        estimateGas(contract=payable_tester_contract,
```

Comment:
I would focus on why this is failing. Many of the other functions will fail if this fails, because calls like `buildTransaction` will call `estimateGas` under the hood, if no gas limit is supplied (which is common).

Can you reproduce outside of testing, with `payable_tester_contract.estimateGas(dict(value=1)).doNoValueCall()`?

Reply:
I could reproduce it. Will dig deeper into it.
```
>>> contract_instance.estimateGas(dict(value=1)).doNoValueCall()
API Exception: {'type': 'TransactionFailed', 'args': (), 'message': ''}
Traceback (most recent call last):
  File "/Users/ankit/web3.py/venv/lib/python3.7/site-packages/jsonrpc/manager.py", line 112, in _get_responses
  File "/Users/ankit/web3.py/venv/lib/python3.7/site-packages/testrpc/server.py", line 34, in inner
  File "/Users/ankit/web3.py/venv/lib/python3.7/site-packages/testrpc/rpc.py", line 138, in eth_estimateGas
  File "/Users/ankit/web3.py/venv/lib/python3.7/site-packages/testrpc/client/client.py", line 253, in estimate_gas
  File "/Users/ankit/web3.py/venv/lib/python3.7/site-packages/testrpc/client/client.py", line 263, in send_transaction
  File "/Users/ankit/web3.py/venv/lib/python3.7/site-packages/testrpc/client/client.py", line 58, in inner
  File "/Users/ankit/web3.py/venv/lib/python3.7/site-packages/testrpc/client/utils.py", line 104, in inner
  File "/Users/ankit/web3.py/venv/lib/python3.7/site-packages/testrpc/client/client.py", line 225, in _send_transaction
  File "/Users/ankit/web3.py/venv/lib/python3.7/site-packages/ethereum/tester.py", line 338, in send
  File "/Users/ankit/web3.py/venv/lib/python3.7/site-packages/ethereum/tester.py", line 296, in _send
ethereum.tester.TransactionFailed
127.0.0.1 - - [13/Sep/2018 08:08:42] "POST / HTTP/1.1" 200 147
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/ankit/web3.py/web3/contract.py", line 1506, in estimate_gas_for_function
    gas_estimate = web3.eth.estimateGas(estimate_transaction)
  File "/Users/ankit/web3.py/web3/eth.py", line 304, in estimateGas
    [transaction],
  File "/Users/ankit/web3.py/web3/manager.py", line 112, in request_blocking
    raise ValueError(response["error"])
ValueError: {'code': -32000, 'message': 'Server error', 'data': {'type': 'TransactionFailed', 'args': [], 'message': ''}}
```